walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,208 @@
 
 package com.google.sps;
 
+import java.lang.Math;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // Get request duration
+    int requestDuration = (int) request.getDuration();
+
+    // If requested duration is longer than one day, return no options as a valid time range.
+    if (requestDuration > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    // Get mandatory attendees
+    Collection<String> attendees = request.getAttendees();
+
+    // Get optional attendees
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+    // If no mandatory or optional attendees are requested, return the whole day as a valid time range.
+    if (attendees.size() == 0 && optionalAttendees.size() == 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // If only mandatory attendees are requested, return available time ranges for mandatory attendees.
+    if (optionalAttendees.size() == 0) {
+      ArrayList<TimeRange> availableTimeRanges = returnAvailableTimeRanges(events, attendees, requestDuration);
+      return availableTimeRanges;
+    }
+
+    // If only optional attendees are requested, return available time ranges for optional attendees.
+    if (attendees.size() == 0) {
+      ArrayList<TimeRange> availableTimeRanges = returnAvailableTimeRanges(events, optionalAttendees, requestDuration);
+      return availableTimeRanges;
+    }
+
+    // Else: Assume that both mandatory and optional attendees are requested.
+    ArrayList<TimeRange> mandatoryAvailableTimeRanges = returnAvailableTimeRanges(events, attendees, requestDuration);
+    ArrayList<TimeRange> optionalAvailableTimeRanges = returnAvailableTimeRanges(events, optionalAttendees, requestDuration);
+    ArrayList<TimeRange> mutualAvailableTimeRanges = getMutualAvailableTimeRanges(mandatoryAvailableTimeRanges, optionalAvailableTimeRanges, requestDuration);
```

Comment:
nit: it's good practice to keep no more than 80 characters in one line

Reply:
Fixed